index.bs

```diff
@@ -485,6 +520,150 @@ spec:csp3; type:grammar; text:base64-value
+
+  1. Let |headers| be |response|'s <a>header list</a>.
+  1. If |headers| <a for="header list">contains</a> a <a>header</a>
+     whose <a>header name</a> is "Integrity-Policy",
+     let |container|'s <a>integrity policy</a> be the result of running
+     <a>processing an integrity policy</a> with the corresponding <a>header value</a>.
+  1. If |headers| <a for="header list">contains</a> a <a>header</a>
+     whose <a>header name</a> is "Integrity-Policy-Report-Only",
+     let |container|'s <a>report only integrity policy</a> be the result of running
+     <a>processing an integrity policy</a> with the corresponding <a>header value</a>.
+  
+  ### Should request be blocked by Integrity Policy ### {#should-request-be-blocked-by-integrity-policy-section}
+  To determine <dfn>should request be blocked by integrity policy</dfn>, given a <a>request</a> |request|,
+  do the following:
+
+  1. Let |policy container| be |request|'s <a>policy container</a>.
+  1. if |request|'s <a>integrity metadata</a> is not the empty string
+     and |request|'s <a for=request>mode</a> is either "cors" or "same-origin",
+     return "Allowed".
+  1. Let |policy| be |policy container|'s <a>integrity policy</a>.
+  1. Let |report policy| be |policy container|'s <a>report only integrity policy</a>.
+  1. Let |block| be a boolean, initially false.
+  1. Let |report block| be a boolean, initially false.
+  1. if |policy|'s <a>sources</a> <a for=list>contains</a> "inline"
+     and |policy|'s <a>blocked destinations</a> <a for=list>contains</a>
+     |request|'s <a for=request>destination</a>,
+     set |block| to true.
+  1. if |policy|'s <a>sources</a> <a for=list>contains</a> "inline"
+     and |report policy|'s <a>blocked destinations</a> <a for=list>contains</a>
+     |request|'s <a for=request>destination</a>,
+     set |report block| to true.
+  1. If |block| or |report block|, <a>report violation</a> with |request|, |block|, |report block|,
+     |policy| and |report policy|.
+  1. If |block|, return "Blocked". Otherwise, return "Allowed".
+ 
+  ### Report violations ### {#report-violations}
+
+  <pre class="idl">
+    [Exposed=Window]
+    interface IntegrityPolicyViolationReportBody : ReportBody {
+      [Default] object toJSON();
+      readonly attribute USVString documentURL;
+      readonly attribute USVString blockedURL;
+      readonly attribute USVString destination;
+      readonly attribute boolean   reportOnly;
+    };
+  </pre>
+
+  To <dfn>report violation</dfn> given a <a>Request</a> |request|, a boolean |block|,
+  a boolean |report block|, an <a>integrity policy struct</a> |policy|,
+  and an <a>integrity policy struct</a> |report policy|, do the following:
+
+  1. Assert that |request|'s <a for=request>client</a> is not null.
+  1. Let |settings object| be |request|'s <a for=request>client</a>.
+  1. Let |global| be |settings object|'s <a>global object</a>.
+  1. Assert that |global| is a <a>Window</a>.
```

Comment:
What would happen if the script was loaded from a worker (e.g., `importScripts`)?

Reply:
Good point. I'll make the exclusion of worker requests from the relevant checks more explicit.

Reply:
Pardon me, why do we _want_ to exclude worker requests? I was under the impression that we are intentionally using request destinations such that we can eventually build an exhaustive mechanism that spans all requests and subresources.

Reply:
I think we want to exclude them (for now) because we don't currently have a reasonable way to currently provide them with integrity metadata. `importScripts` just takes in URLs, worker scripts [options](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker#options) don't currently have an `integrity` attribute, and module worker scripts could have benefitted from [import map integrity](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script/type/importmap#integrity_metadata_map) if they [supported import maps at all](https://github.com/whatwg/html/pull/10858).

Once we'd have a reasonable way to pass inline metadata to them, we can enforce that using the `worker`/`sharedworker`/`serviceworker` destinations.

Reply:
Given that all worker environments can have a CSP I would expect them to also be able to have an Integrity Policy. Perhaps if you specify one now that means `importScripts()` won't work, but `fetch()` still does. But I don't think `importScripts()` not working is sufficient reason to not make it work at all. As then we might end up in a situation where servers start sending this header for workers and expect it to be ignored.

Reply:
I guess that specifying Integrity-Policy for script destinations in worker environments now would mean that Service Workers can't `fetch()` Requests for scripts without integrity.

Reply:
And the same for dedicated and shared workers, yes. Seems good?

Reply:
Hmm, I'm probably missing something, but can a `fetch()` call set the Request destination to be "script", if it didn't get the Request object from the browser? (e.g. through a SW "fetch" event)

Reply:
Oh right, this is only for script destinations. In that case you're probably correct, although I'm not sure what the state of import maps for workers is and such. In any event I think we should support this in the same places we would have supported it if it were part of CSP.

Reply:
OK, added the required handling to handle workers as well.